craft_application/services/lifecycle.py

```diff
@@ -268,6 +282,22 @@ def __repr__(self) -> str:
             f"{work_dir=}, {cache_dir=}, {build_for=}, **{self._manager_kwargs!r})"
         )
 
+    def _update_project_metadata(self) -> None:
+        """Replace project fields with values adopted during the lifecycle."""
+        self._update_project_variables()
+
+    def _update_project_variables(self) -> None:
+        """Replace project fields with values set using craftctl."""
+        update_vars: dict[str, str] = {}
+        for var in self._app.project_variables:
+            value = self._lcm.project_info.get_project_var(var)
```

Comment:
you can also do `self.project_info` (it's a property)

Reply:
Updated to use `self._project_info`.